In [88]:
import pandas as pd

In [89]:
movies = pd.read_csv('/content/drive/MyDrive/Data Science/dataset/imdb-top-1000.csv')

**1. Shape of the DataSet**

In [90]:
movies.shape

(1000, 10)

**2. First 10 rows of the DataSet**

In [91]:
movies.head(10)

Series_Title Released_Year  Runtime  \
0                       The Shawshank Redemption          1994      142   
1                                  The Godfather          1972      175   
2                                The Dark Knight          2008      152   
3                         The Godfather: Part II          1974      202   
4                                   12 Angry Men          1957       96   
5  The Lord of the Rings: The Return of the King          2003      201   
6                                   Pulp Fiction          1994      154   
7                               Schindler's List          1993      195   
8                                      Inception          2010      148   
9                                     Fight Club          1999      139   

       Genre  IMDB_Rating              Director              Star1  \
0      Drama          9.3        Frank Darabont        Tim Robbins   
1      Crime          9.2  Francis Ford Coppola      Marlon Brando   
2     Action          9.0     Christopher Nolan     Christian Bale   
3      Crime          9.0  Francis Ford Coppola          Al Pacino   
4      Crime          9.0          Sidney Lumet        Henry Fonda   
5     Action          8.9         Peter Jackson        Elijah Wood   
6      Crime          8.9     Quentin Tarantino      John Travolta   
7  Biography          8.9      Steven Spielberg        Liam Neeson   
8     Action          8.8     Christopher Nolan  Leonardo DiCaprio   
9      Drama          8.8         David Fincher          Brad Pitt   

   No_of_Votes        Gross  Metascore  
0      2343110   28341469.0       80.0  
1      1620367  134966411.0      100.0  
2      2303232  534858444.0       84.0  
3      1129952   57300000.0       90.0  
4       689845    4360000.0       96.0  
5      1642758  377845905.0       94.0  
6      1826188  107928762.0       94.0  
7      1213505   96898818.0       94.0  
8      2067042  292576195.0       74.0  
9      1854740   37030102.0       66.0

**3. Random 10 rows of the DataSet**

In [92]:
movies.sample(10)

Series_Title Released_Year  Runtime  \
148                                Chak De! India          2007      153   
589                            La grande bellezza          2013      141   
36                                   The Prestige          2006      130   
729                      The Theory of Everything          2014      123   
982                                     Mad Max 2          1981       96   
226  Harry Potter and the Deathly Hallows: Part 2          2011      130   
279                                         Rocky          1976      120   
344               The Perks of Being a Wallflower          2012      103   
140                            Gangs of Wasseypur          2012      321   
156                        Swades: We, the People          2004      210   

         Genre  IMDB_Rating            Director               Star1  \
148      Drama          8.2         Shimit Amin      Shah Rukh Khan   
589      Drama          7.8    Paolo Sorrentino       Toni Servillo   
36       Drama          8.5   Christopher Nolan      Christian Bale   
729  Biography          7.7         James Marsh      Eddie Redmayne   
982     Action          7.6       George Miller          Mel Gibson   
226  Adventure          8.1         David Yates    Daniel Radcliffe   
279      Drama          8.1    John G. Avildsen  Sylvester Stallone   
344      Drama          8.0     Stephen Chbosky        Logan Lerman   
140     Action          8.2      Anurag Kashyap      Manoj Bajpayee   
156      Drama          8.2  Ashutosh Gowariker      Shah Rukh Khan   

     No_of_Votes        Gross  Metascore  
148        74129    1113541.0       68.0  
589        81125    2852400.0       86.0  
36       1190259   53089891.0       66.0  
729       404182   35893537.0       72.0  
982       166588   12465371.0       77.0  
226       764493  381011219.0       85.0  
279       518546  117235247.0       70.0  
344       462252   17738570.0       67.0  
140        82365    5257741.0       89.0  
156        83005    1223240.0        NaN

**4. Lasst 10 rows of the DataSet**

In [93]:
movies.tail(10)

Series_Title Released_Year  Runtime      Genre  IMDB_Rating  \
990            Giù la testa          1971      157      Drama          7.6   
991          Kelly's Heroes          1970      144  Adventure          7.6   
992         The Jungle Book          1967       78  Animation          7.6   
993                  Blowup          1966      111      Drama          7.6   
994      A Hard Day's Night          1964       87     Comedy          7.6   
995  Breakfast at Tiffany's          1961      115     Comedy          7.6   
996                   Giant          1956      201      Drama          7.6   
997   From Here to Eternity          1953      118      Drama          7.6   
998                Lifeboat          1944       97      Drama          7.6   
999            The 39 Steps          1935       86      Crime          7.6   

                   Director              Star1  No_of_Votes        Gross  \
990            Sergio Leone        Rod Steiger        30144     696690.0   
991         Brian G. Hutton     Clint Eastwood        45338    1378435.0   
992     Wolfgang Reitherman        Phil Harris       166409  141843612.0   
993  Michelangelo Antonioni     David Hemmings        56513  632532802.0   
994          Richard Lester        John Lennon        40351   13780024.0   
995           Blake Edwards     Audrey Hepburn       166544  679874270.0   
996          George Stevens   Elizabeth Taylor        34075  195217415.0   
997          Fred Zinnemann     Burt Lancaster        43374   30500000.0   
998        Alfred Hitchcock  Tallulah Bankhead        26471  852142728.0   
999        Alfred Hitchcock       Robert Donat        51853  302787539.0   

     Metascore  
990       77.0  
991       50.0  
992       65.0  
993       82.0  
994       96.0  
995       76.0  
996       84.0  
997       85.0  
998       78.0  
999       93.0

**5. Information about the columns**

In [94]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Series_Title   1000 non-null   object 
 1   Released_Year  1000 non-null   object 
 2   Runtime        1000 non-null   int64  
 3   Genre          1000 non-null   object 
 4   IMDB_Rating    1000 non-null   float64
 5   Director       1000 non-null   object 
 6   Star1          1000 non-null   object 
 7   No_of_Votes    1000 non-null   int64  
 8   Gross          1000 non-null   float64
 9   Metascore      843 non-null    float64
dtypes: float64(3), int64(2), object(5)
memory usage: 78.2+ KB


**Understandings:** There are 10 columns. The "Series_Title", "Released_Year", "Genre", "Derector" and "Star1" column contain string data. On the other hand, "Runtime" and "No_of_Votes" contain int data. Lastly, "IMDB_Rating", "Gross" and "Metascore" contain float data. However, only "Metascore" column contains 157(1000-843) Null values.

**6. Check duplicate entity**

In [95]:
movies.duplicated().sum()

0

There is no duplicate entity.

**7. Insights about numerical column**

In [96]:
movies.describe()

Runtime  IMDB_Rating   No_of_Votes         Gross   Metascore
count  1000.000000  1000.000000  1.000000e+03  1.000000e+03  843.000000
mean    122.891000     7.949300  2.736929e+05  1.282781e+08   77.971530
std      28.093671     0.275491  3.273727e+05  2.011110e+08   12.376099
min      45.000000     7.600000  2.508800e+04  1.305000e+03   28.000000
25%     103.000000     7.700000  5.552625e+04  5.012919e+06   70.000000
50%     119.000000     7.900000  1.385485e+05  3.953432e+07   79.000000
75%     137.000000     8.100000  3.741612e+05  1.615997e+08   87.000000
max     321.000000     9.300000  2.343110e+06  9.366622e+08  100.000000

**8. Find the movie(s) with maximum rating.**

In [97]:
mask = movies['IMDB_Rating'] == movies['IMDB_Rating'].max()
movies[mask]

Series_Title Released_Year  Runtime  Genre  IMDB_Rating  \
0  The Shawshank Redemption          1994      142  Drama          9.3   

         Director        Star1  No_of_Votes       Gross  Metascore  
0  Frank Darabont  Tim Robbins      2343110  28341469.0       80.0

**9. Find the movie(s) with minimum rating.**

In [98]:
mask = movies['IMDB_Rating'] == movies['IMDB_Rating'].min()
movies[mask]

Series_Title Released_Year  Runtime      Genre  \
877                       Dark Waters          2019      126  Biography   
878                         Searching          2018      102      Drama   
879  Once Upon a Time... in Hollywood          2019      161     Comedy   
880                          Nelyubov          2017      127      Drama   
881               The Florida Project          2017      111      Drama   
..                                ...           ...      ...        ...   
995            Breakfast at Tiffany's          1961      115     Comedy   
996                             Giant          1956      201      Drama   
997             From Here to Eternity          1953      118      Drama   
998                          Lifeboat          1944       97      Drama   
999                      The 39 Steps          1935       86      Crime   

     IMDB_Rating            Director              Star1  No_of_Votes  \
877          7.6         Todd Haynes       Mark Ruffalo        60408   
878          7.6     Aneesh Chaganty           John Cho       140840   
879          7.6   Quentin Tarantino  Leonardo DiCaprio       551309   
880          7.6  Andrey Zvyagintsev     Maryana Spivak        29765   
881          7.6          Sean Baker   Brooklynn Prince        95181   
..           ...                 ...                ...          ...   
995          7.6       Blake Edwards     Audrey Hepburn       166544   
996          7.6      George Stevens   Elizabeth Taylor        34075   
997          7.6      Fred Zinnemann     Burt Lancaster        43374   
998          7.6    Alfred Hitchcock  Tallulah Bankhead        26471   
999          7.6    Alfred Hitchcock       Robert Donat        51853   

           Gross  Metascore  
877  428167124.0       73.0  
878   26020957.0       71.0  
879  142502728.0       83.0  
880     566356.0       86.0  
881    5904366.0       92.0  
..           ...        ...  
995  679874270.0       76.0  
996  195217415.0       84.0  
997   30500000.0       85.0  
998  852142728.0       78.0  
999  302787539.0       93.0  

[123 rows x 10 columns]

**10. Find the movie(s) with maximum gross.**

In [99]:
mask = movies['Gross'] == movies['Gross'].max()
movies[mask]

Series_Title Released_Year  Runtime  \
477  Star Wars: Episode VII - The Force Awakens          2015      138   

      Genre  IMDB_Rating     Director         Star1  No_of_Votes        Gross  \
477  Action          7.9  J.J. Abrams  Daisy Ridley       860823  936662225.0   

     Metascore  
477       80.0

**11. Find the movie(s) with minimum gross.**

In [100]:
mask = movies['Gross'] == movies['Gross'].min()
movies[mask]

Series_Title Released_Year  Runtime   Genre  IMDB_Rating  \
630  Adams æbler          2005       94  Comedy          7.8   

                 Director           Star1  No_of_Votes   Gross  Metascore  
630  Anders Thomas Jensen  Ulrich Thomsen        45717  1305.0       51.0

**12. Which movie(s) earn most with lowest rating?**

In [101]:
mask = movies['IMDB_Rating'] == movies['IMDB_Rating'].min()
movies[mask].sort_values("Gross", ascending = False).head(1)

Series_Title Released_Year  Runtime  Genre  IMDB_Rating       Director  \
913    Die Welle          2008      107  Drama          7.6  Dennis Gansel   

            Star1  No_of_Votes        Gross  Metascore  
913  Jürgen Vogel       102742  924558264.0        NaN

**13. Which is the longest movie(s)?**

In [102]:
mask = movies['Runtime'] == movies['Runtime'].max()
movies[mask]

Series_Title Released_Year  Runtime   Genre  IMDB_Rating  \
140  Gangs of Wasseypur          2012      321  Action          8.2   

           Director           Star1  No_of_Votes      Gross  Metascore  
140  Anurag Kashyap  Manoj Bajpayee        82365  5257741.0       89.0

**14. Which is the shortest movie(s)?**

In [103]:
mask = movies['Runtime'] == movies['Runtime'].min()
movies[mask]

Series_Title Released_Year  Runtime   Genre  IMDB_Rating       Director  \
194  Sherlock Jr.          1924       45  Action          8.2  Buster Keaton   

             Star1  No_of_Votes     Gross  Metascore  
194  Buster Keaton        41985  977375.0        NaN

**15. Find the top 3 genres by total earning.**

In [104]:
movies.groupby("Genre")["Gross"].sum().sort_values(ascending = False).head(3).reset_index()

Genre         Gross
0   Drama  3.540997e+10
1  Action  3.263226e+10
2  Comedy  1.566387e+10

**16. Write a function which takes genre as argument and return average gross and total gross.**

In [105]:
def give_gross(genre):
    mask = movies['Genre'] == genre
    avg_gross = movies[mask]['Gross'].mean()
    total_gross = movies[mask]['Gross'].sum()

    return avg_gross, total_gross

genre = "Action"
avg, total = give_gross(genre)

print(f"Genre: {genre}\nAverage Gross: {avg}\nTotal Gross: {total}")

Genre: Action
Average Gross: 189722449.5
Total Gross: 32632261314.0


**17. Find the genre with highest average IMDB rating.**

In [106]:
df = movies.groupby('Genre')['IMDB_Rating'].mean().sort_values(ascending = False).head(1).reset_index()
df = df.rename(columns = {"IMDB_Rating": "Average Rating"})
df

Genre  Average Rating
0  Western            8.35

**18. Find director with most popularity.**

In [107]:
movies.groupby("Director")['No_of_Votes'].sum().sort_values(ascending = False).head(1).index[0]

'Christopher Nolan'

**19. Find the most frequent Director.**

In [108]:
# Approach - 1

df = movies['Director'].value_counts().sort_values().tail(1).reset_index()
df = df.rename(columns = {"count": "Movie Count"})
df

Director  Movie Count
0  Alfred Hitchcock           14

In [109]:
# Approach - 2

df = movies.groupby('Director')['Series_Title'].count().sort_values().tail(1).reset_index()
df = df.rename(columns = {"Series_Title": "Movie Count"})
df

Director  Movie Count
0  Alfred Hitchcock           14

**20. Find the highest rated movie of each genre.**

In [110]:
# Approach - 1

series_of_max_rating = movies.groupby('Genre')['IMDB_Rating'].max()
genres = list(series_of_max_rating.index)
max_ratings = list(series_of_max_rating.values)

ans = pd.DataFrame()

for genre, max_rating in zip(genres, max_ratings):
    mask = ((movies['Genre'] == genre) & (movies['IMDB_Rating'] == max_rating))
    ans = pd.concat([ans, movies[mask]], ignore_index = True)

ans[['Genre', 'Series_Title', 'IMDB_Rating']]

Genre                         Series_Title  IMDB_Rating
0      Action                      The Dark Knight          9.0
1   Adventure                         Interstellar          8.6
2   Animation        Sen to Chihiro no kamikakushi          8.6
3   Biography                     Schindler's List          8.9
4      Comedy                         Gisaengchung          8.6
5      Comedy                      La vita è bella          8.6
6       Crime                        The Godfather          9.2
7       Drama             The Shawshank Redemption          9.3
8      Family           E.T. the Extra-Terrestrial          7.8
9      Family  Willy Wonka & the Chocolate Factory          7.8
10    Fantasy         Das Cabinet des Dr. Caligari          8.1
11  Film-Noir                        The Third Man          8.1
12     Horror                               Psycho          8.5
13    Mystery                              Memento          8.4
14    Mystery                          Rear Window          8.4
15   Thriller                      Wait Until Dark          7.8
16    Western      Il buono, il brutto, il cattivo          8.8

In [111]:
# Approach - 2: Looping over Groups

genres = movies.groupby('Genre')

for genre, df in genres:
    mask = df["IMDB_Rating"] == df['IMDB_Rating'].max()
    print(genre, ": ", list(df[mask]["Series_Title"]))

Action :  ['The Dark Knight']
Adventure :  ['Interstellar']
Animation :  ['Sen to Chihiro no kamikakushi']
Biography :  ["Schindler's List"]
Comedy :  ['Gisaengchung', 'La vita è bella']
Crime :  ['The Godfather']
Drama :  ['The Shawshank Redemption']
Family :  ['E.T. the Extra-Terrestrial', 'Willy Wonka & the Chocolate Factory']
Fantasy :  ['Das Cabinet des Dr. Caligari']
Film-Noir :  ['The Third Man']
Horror :  ['Psycho']
Mystery :  ['Memento', 'Rear Window']
Thriller :  ['Wait Until Dark']
Western :  ['Il buono, il brutto, il cattivo']


In [112]:
# Approach - 3: Using Apply

def max_rated(genre):
    global ans
    mask = genre["IMDB_Rating"] == genre['IMDB_Rating'].max()
    ans = pd.concat([ans, genre[mask]], ignore_index = True)


genres = movies.groupby('Genre')
ans = pd.DataFrame()
genres.apply(max_rated)

ans[['Genre', 'Series_Title', 'IMDB_Rating']]

Genre                         Series_Title  IMDB_Rating
0      Action                      The Dark Knight          9.0
1   Adventure                         Interstellar          8.6
2   Animation        Sen to Chihiro no kamikakushi          8.6
3   Biography                     Schindler's List          8.9
4      Comedy                         Gisaengchung          8.6
5      Comedy                      La vita è bella          8.6
6       Crime                        The Godfather          9.2
7       Drama             The Shawshank Redemption          9.3
8      Family           E.T. the Extra-Terrestrial          7.8
9      Family  Willy Wonka & the Chocolate Factory          7.8
10    Fantasy         Das Cabinet des Dr. Caligari          8.1
11  Film-Noir                        The Third Man          8.1
12     Horror                               Psycho          8.5
13    Mystery                              Memento          8.4
14    Mystery                          Rear Window          8.4
15   Thriller                      Wait Until Dark          7.8
16    Western      Il buono, il brutto, il cattivo          8.8

**21. Write a function which takes genre as aurgument and return highest rated movie of that genre.**

In [114]:
def max_rated_movie(genre):
    mask = movies["Genre"] == genre
    df = movies[mask]

    return df[df["IMDB_Rating"] == df["IMDB_Rating"].max()]

max_rated_movie("Action")

Series_Title Released_Year  Runtime   Genre  IMDB_Rating  \
2  The Dark Knight          2008      152  Action          9.0   

            Director           Star1  No_of_Votes        Gross  Metascore  
2  Christopher Nolan  Christian Bale      2303232  534858444.0       84.0

**22. Find number of movies starting with A for each group.**

In [115]:
def movie_count(genre):
    return genre["Series_Title"].str.startswith('A').sum()

genres = movies.groupby('Genre')
genres.apply(movie_count)

Genre
Action       10
Adventure     2
Animation     2
Biography     9
Comedy       14
Crime         4
Drama        21
Family        0
Fantasy       0
Film-Noir     0
Horror        1
Mystery       0
Thriller      0
Western       0
dtype: int64

**23. Find ranking of each movie in each genre according to IMDB score. Rank 1 goes to the highest rated movie, 2 goes to the 2nd highest rated movie and so on.**

In [116]:
def find_rank(genre):
    genre["Rank"] = genre["IMDB_Rating"].rank(ascending = False)
    return genre

genres = movies.groupby('Genre')
genres.apply(find_rank)

Series_Title Released_Year  \
Genre                                                                           
Action   2                                      The Dark Knight          2008   
         5        The Lord of the Rings: The Return of the King          2003   
         8                                            Inception          2010   
         10   The Lord of the Rings: The Fellowship of the Ring          2001   
         13               The Lord of the Rings: The Two Towers          2002   
...                                                         ...           ...   
Thriller 700                                    Wait Until Dark          1967   
Western  12                     Il buono, il brutto, il cattivo          1966   
         48                        Once Upon a Time in the West          1968   
         115                         Per qualche dollaro in più          1965   
         691                             The Outlaw Josey Wales          1976   

              Runtime     Genre  IMDB_Rating           Director  \
Genre                                                             
Action   2        152    Action          9.0  Christopher Nolan   
         5        201    Action          8.9      Peter Jackson   
         8        148    Action          8.8  Christopher Nolan   
         10       178    Action          8.8      Peter Jackson   
         13       179    Action          8.7      Peter Jackson   
...               ...       ...          ...                ...   
Thriller 700      108  Thriller          7.8      Terence Young   
Western  12       161   Western          8.8       Sergio Leone   
         48       165   Western          8.5       Sergio Leone   
         115      132   Western          8.3       Sergio Leone   
         691      135   Western          7.8     Clint Eastwood   

                          Star1  No_of_Votes        Gross  Metascore  Rank  
Genre                                                                       
Action   2       Christian Bale      2303232  534858444.0       84.0   1.0  
         5          Elijah Wood      1642758  377845905.0       94.0   2.0  
         8    Leonardo DiCaprio      2067042  292576195.0       74.0   3.5  
         10         Elijah Wood      1661481  315544750.0       92.0   3.5  
         13         Elijah Wood      1485555  342551365.0       87.0   6.0  
...                         ...          ...          ...        ...   ...  
Thriller 700     Audrey Hepburn        27733   17550741.0       81.0   1.0  
Western  12      Clint Eastwood       688390    6100000.0       90.0   1.0  
         48         Henry Fonda       302844    5321508.0       80.0   2.0  
         115     Clint Eastwood       232772   15000000.0       74.0   3.0  
         691     Clint Eastwood        65659   31800000.0       69.0   4.0  

[1000 rows x 11 columns]

**24.  Find the most earning actor->director combo.**

In [124]:
pair = movies.groupby(['Star1', 'Director'])
pair["Gross"].sum().sort_values(ascending = False).head(1).reset_index()

Star1        Director         Gross
0  Toshirô Mifune  Akira Kurosawa  2.999877e+09

**25. Find the best(in-terms of metascore(avg)) actor->genre combo.**

In [126]:
pair = movies.groupby(['Star1', 'Genre'])
pair["Metascore"].mean().sort_values(ascending = False).head(1).reset_index()

Star1  Genre  Metascore
0  Ellar Coltrane  Drama      100.0

**26. Find the maximum movie release year.**

In [134]:
# Approach - 1

df = movies.groupby('Released_Year').size().sort_values().tail(1).reset_index()
df = df.rename(columns = {0: 'Count'})
df

Released_Year  Count
0          2014     32

In [136]:
# Approach - 2

movies['Released_Year'].value_counts().sort_values().tail(1).reset_index()

Released_Year  count
0          2014     32